File: src/main/java/fr/ill/ics/cameo/impl/ResponderImpl.java
/*
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under the EUPL, Version 1.1 only (the "License");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * http://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */

package fr.ill.ics.cameo.impl;

import org.zeromq.ZContext;
import org.zeromq.ZMQ;
import org.zeromq.ZMQ.Socket;
import org.zeromq.ZMsg;

import com.google.protobuf.InvalidProtocolBufferException;

import fr.ill.ics.cameo.proto.Messages.MessageType;
import fr.ill.ics.cameo.proto.Messages.MessageType.Type;

public class ResponderImpl {

	public static final String RESPONDER_PREFIX = "rep.";

	private ApplicationImpl application;
	ZContext context;
	private int responderPort;
	private String name;
	private Socket responder = null;
	private boolean ended = false;
	private ResponderWaitingImpl waiting = new ResponderWaitingImpl(this);
	
	public ResponderImpl(ApplicationImpl application, ZContext context, int responderPort, String name) {
		this.application = application;
		this.context = context;
		this.responderPort = responderPort;
		this.name = name;

		// create a socket REP
		responder = context.createSocket(ZMQ.REP);
		responder.bind("tcp://*:" + responderPort);
		
		waiting.add();
	}
	
	public String getName() {
		return name;
	}

	public RequestImpl receive() {
		
		ZMsg message = null;
		ZMsg reply = null;
		
		try {
			message = ZMsg.recvMsg(responder);

			if (message == null) {
				ended = true;
				
				return null;
			}

			// Check there are not 2 frames
			if (message.size() != 2) {
				System.err.println("Unexpected number of frames, should be 2");
				ended = true;
				
				return null;
			}
			// 2 frames, get first frame (type)
			byte[] typeData = message.getFirst().getData();
			// Get last frame
			byte[] messageData = message.getLast().getData();
			
			// dispatch message
			MessageType type = MessageType.parseFrom(typeData);
			
			if (type.getType() == Type.REQUEST) {
				// Parse the message
				fr.ill.ics.cameo.proto.Messages.Request request = fr.ill.ics.cameo.proto.Messages.Request.parseFrom(messageData);
			
				// Create the request
				return new RequestImpl(application, context, request.getEndpoint(), request.getMessage(), request.getApplicationId());
				
			} else if (type.getType() == Type.CANCEL) {
				ended = true;
				
				return null;
			}
			
		} catch (InvalidProtocolBufferException e) {
			System.err.println("problem in parsing of message");
			
		} finally {
			
			if (message != null) {
				message.destroy();
			}	

			// Send to the requester
			reply = new ZMsg();
			reply.add("OK");
			reply.send(responder);
			
			if (reply != null) {
				reply.destroy();
			}
		}
			
		return null;
	}
	
	public void cancel() {
		String endpoint = application.getUrl() + ":" + responderPort;

		ZMsg requestMessage = application.createRequest(Type.CANCEL);
		String content = "cancel";
		requestMessage.add(content);
		
		application.tryRequest(requestMessage, endpoint);
	}

	public boolean hasEnded() {
		return ended;
	}
	
	public void terminate() {

		waiting.remove();
		context.destroySocket(responder);
		
		try {
			application.removePort(RESPONDER_PREFIX + name);
			
		} catch (Exception e) {
			System.err.println("Cannot terminate responder: " + e.getMessage());
		}
	}
	
	@Override
	public String toString() {
		return RESPONDER_PREFIX + name + ":" + application.getName() + "." + application.getId() + "@" + application.getEndpoint();
	}
	
}

File: src/cameo/impl/ResponderImpl.cpp
/*
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under the EUPL, Version 1.1 only (the "License");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * http://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */

#include "ResponderImpl.h"

#include <boost/bind.hpp>
#include <sstream>
#include "../Application.h"
#include "../Serializer.h"
#include "ApplicationImpl.h"
#include "RequestImpl.h"

using namespace std;
using namespace boost;

namespace cameo {

const std::string ResponderImpl::RESPONDER_PREFIX = "rep.";

ResponderImpl::ResponderImpl(const application::This * application, int responderPort, const std::string& name) :
	m_application(application),
	m_responderPort(responderPort),
	m_name(name),
	m_ended(false) {

	// create a socket REP
	m_responder.reset(new zmq::socket_t(m_application->m_impl->m_context, ZMQ_REP));
	stringstream repEndpoint;
	repEndpoint << "tcp://*:" << m_responderPort;

	m_responder->bind(repEndpoint.str().c_str());
}

ResponderImpl::~ResponderImpl() {
	terminate();
}

void ResponderImpl::cancel() {

	stringstream endpoint;
	endpoint << m_application->getUrl() << ":" << m_responderPort;

	string strRequestType = m_application->m_impl->createRequest(PROTO_CANCEL);
	string strRequestData = "cancel";

	unique_ptr<zmq::message_t> reply = m_application->m_impl->tryRequestWithOnePartReply(strRequestType, strRequestData, endpoint.str());

	proto::RequestResponse requestResponse;
	requestResponse.ParseFromArray((*reply).data(), (*reply).size());
}

WaitingImpl * ResponderImpl::waiting() {
	return new GenericWaitingImpl(bind(&ResponderImpl::cancel, this));
}

std::unique_ptr<RequestImpl> ResponderImpl::receive() {

	zmq::message_t * message = new zmq::message_t;
	m_responder->recv(message, 0);

	// multi-part message, first part is the type
	proto::MessageType messageType;
	messageType.ParseFromArray((*message).data(), (*message).size());

	if (message->more()) {
		delete message;
		message = new zmq::message_t;
		m_responder->recv(message, 0);

	} else {
		cerr << "unexpected number of frames, should be 2" << endl;
		m_ended = true;

		return unique_ptr<RequestImpl>(nullptr);
	}

	// Create the reply
	string data = "OK";
	size_t size = data.length();
	zmq::message_t * reply = new zmq::message_t(size);
	memcpy((void *) reply->data(), data.c_str(), size);

	unique_ptr<RequestImpl> result;

	if (messageType.type() == proto::MessageType_Type_REQUEST) {

		// Parse the message
		proto::Request messageRequest;
		messageRequest.ParseFromArray((*message).data(), (*message).size());

		// Create the request
		result = unique_ptr<RequestImpl>(new RequestImpl(m_application,
				messageRequest.applicationname(),
				messageRequest.applicationid(),
				messageRequest.message(),
				messageRequest.serverurl(),
				messageRequest.serverport(),
				messageRequest.requesterport()));

		// Set message 2 if it exists.
		if (messageRequest.has_message2()) {
			result->m_message2 = messageRequest.message2();
		}

	} else if (messageType.type() == proto::MessageType_Type_CANCEL) {
		m_ended = true;

	} else {
		cerr << "unknown message type " << messageType.type() << endl;
		m_responder->send(*message);
	}

	// send to the client
	if (reply != 0) {
		m_responder->send(*reply);
	}

	delete reply;
	delete message;

	return result;
}

void ResponderImpl::terminate() {

	if (m_responder.get() != nullptr) {
		m_responder.reset(nullptr);

		bool success = m_application->removePort(RESPONDER_PREFIX + m_name);
		if (!success) {
			cerr << "server cannot destroy responder " << m_name << endl;
		}
	}
}

}
